harness/determined/common/util.py

```diff
@@ -207,6 +208,9 @@ def parse_protobuf_timestamp(ts: str) -> datetime.datetime:
     # [2] https://bugs.python.org/issue35829
     if ts.endswith("Z"):
         ts = ts[:-1] + "+00:00"
+    # Remove the [milli,micro,nano]seconds portion from the timestamp because
+    # fromisoformat cannot process nanoseconds and we do not use other values either.
+    re.sub("\\.[0-9]*\\+", "+", ts)
```

Comment:
Hm, interesting.

This is a difference between rfc3339 format and what fromisoformat allows.

fromisoformat allows 0, 3, or 6 decimal digits of seconds ([code here](https://github.com/python/cpython/blob/893c3b7f5c52b5c0800b72a78fb799632efcd628/Modules/_datetimemodule.c#L762)).  At least, that was until python 3.11 was release, at which point [arbitrary length fractional seconds were allowed](https://github.com/python/cpython/blob/34ef75d3ef559288900fad008f05b29155eb8b59/Modules/_datetimemodule.c#L867) and even [the Z timezone was allowed](https://github.com/python/cpython/blob/34ef75d3ef559288900fad008f05b29155eb8b59/Modules/_datetimemodule.c#L915) (whoo!).

Anyway, I'm ok with a regex for this, because:
- there is only one point in an RFC3339 timestamp where the `.` is allowed
- we can remove all of this when python 3.10.x goes EOL (admittedly a long time away).

However, can you:
- change the regex to pattern to `r'\.[0-9]*'` and replace with an empty string (`""`)?  Note the use of the raw string `r''` to avoid double backslashes, which is common when writing regexes in python, and also avoid matching the `+` since `-HH:MM` is also allowed in rfc3339.
- clarify your comment to explain precisely what fromisoformat allows and doesn't allow, and mention that this workaround won't be necessary when 3.11 is the oldest python we support
- include a comment explaining why you're regex is safe on RFC3339.

Sorry for being so pedantic here, but parsing standards-compliant strings with any combination of regex or pattern matching can be very fragile.  That's why you're writing this fix in the first place (the `.split(".")` broke), that's why you're modifying my own RFC3339 parser here (I didn't realize the fractional second restrictions), and I want us to avoid having to fix this again in the future.

Also, having explanations as to what versions of python have a shortcoming make it easier for future developers to delete the workarounds when they become obsolete, rather than leaving the workarounds forever out of fear of what might break.

Reply:
Thanks for the detailed comment, Ryan. Updated the code as suggested.